source/mir/random/tinflex/internal/types.d

```diff
@@ -0,0 +1,318 @@
+module mir.random.tinflex.internal.types;
+
+import std.traits: ReturnType, isFloatingPoint;
+
+protected:
+
+/**
+Major data unit of the Tinflex algorithm.
+It is used to store
+- (cached) values of the transformation (and its derivatives)
+- area below the hat and squeeze function
+- linked-list like reference to the right part of the interval (there will always
+be exactly one interval with right = 0)
+*/
+struct IntervalPoint(S)
+    if (isFloatingPoint!S)
+{
+    import mir.random.tinflex.internal.linearfun : LinearFun;
+
+    /// left position of the interval
+    immutable S x;
+
+    /// T_c family of the interval
+    immutable S c;
+
+    /// transformed values
+    immutable S tx, t1x, t2x;
+
+    this (in S tx, in S t1x, in S t2x, in S x, in S c)
+    {
+        this.tx = tx;
+        this.t1x = t1x;
+        this.t2x = t2x;
+        this.x = x;
+        this.c = c;
+        import std.stdio;
+        writeln("CC", c);
+    }
+
+    FunType type;
+
+    LinearFun!S hat;
+    LinearFun!S squeeze;
+
+    // make NaN
+    S hatA = 0;
+    S squeezeA = 0;
+
+    // disallow NaN points
+    invariant {
+        import std.math : isNaN;
+        import std.meta : AliasSeq;
+        import std.traits : isFloatingPoint;
+        static if (isFloatingPoint!S)
+        {
+            alias seq =  AliasSeq!(x, c, tx, t1x, t2x);
+            foreach (i, v; seq)
+            {
+                assert(!v.isNaN, "variable " ~ seq[i].stringof ~ " isn't allowed to be NaN");
+            }
+        }
+    }
+}
+
+/// ditto
+IntervalPoint!S intervalPoint(F0, F1, F2, S)
+                             (in F0 f0, in F1 f1, in F2 f2, in S x, in S c = 1.5)
+    if (is(ReturnType!F0 == S) && is(ReturnType!F1 == S) && is(ReturnType!F2 == S) &&
+        (isFloatingPoint!S))
+{
+    return intervalPoint(f0(x), f1(x), f2(x), x, c);
+}
+
+IntervalPoint!S intervalPoint(S)
+                             (in S f0, in S f1, in S f2, in S x, in S c = 1.5)
+    if (isFloatingPoint!S)
+{
+    return IntervalPoint!S(f0, f1, f2, x, c);
+}
+
+/**
+Notations of different function types according to the Tinflex paper.
+It is based on this naming scheme:
+
+- a: concAve
+- b: convex
+- Type 4 is the pure case without any inflection point
+*/
+enum FunType {T1a, T1b, T2a, T2b, T3a, T3b, T4a, T4b}
+
+/**
+Checks whether the function type is concave
+*/
+bool isConcave(in FunType t) pure @safe @nogc nothrow
+{
+    with(FunType)
+    return t == T1a || t == T2a || t == T3a || t == T4a;
```

Comment:
you forgot to optimise this with my suggestion. please find it in comment for previous commits